### PR TITLE
change: Move from `maybe_rayon` to `rayon-1.8`

### PR DESCRIPTION
**Pull request by CPerezz**
After seeing the issues with WASM that the last release (`0.5.0`) of this crate made, the idea is to now control the compatibility with WASM while at the same time, making it easy to handle.

In this line of work, the idea was to simply do the following:
- Use `rayon 1.8` as a dependency for paralellism which fixes the WASM compat issues with `multicore`-related things. See: https://github.com/rayon-rs/rayon/pull/1019 thanks @han0110 for the suggestion.
- Use conditional dev-dependency loading for `getrandom` such that we can compile the lib for WASM-targets in the CI without needing to have the dependency being pulled downstream.
- The `multicore` module is gone, and the rest of the code has been refactored since the "fallback" is now handled by rayon itself.
- The CI has also been modified to keep accounting for WASM compatibility by forcing the build to `wasmi` and `wasm32` targets on each CI run. (It does so by targeting `test` build such that we can get `getrandom` as a dev-dependency which will not get pulled on consumer's machines.
- Finally, updates the README to provide further info on how this lib should be used by a consumer if they want to use it within WASM.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -42,8 +41,10 @@ jobs:
     strategy:
       matrix:
         include:
-          - feature: 
           - feature: default
+        target:
+          - wasm32-unknown-unknown
+          - wasm32-wasi
```
Comment (han0110) [inline .github/workflows/ci.yml:47]: I think only adding this wouldn't really make cargo build for them, should we also add `--target "${{ matrix.target }}"` in the following commands?

Also I think we could separate the build check for wasm as another job (just like halo2 [test](https://github.com/privacy-scaling-explorations/halo2/blob/main/.github/workflows/ci.yml#L11) and [build](https://github.com/privacy-scaling-explorations/halo2/blob/main/.github/workflows/ci.yml#L35) is separated), that might make the CI faster?

Comment (CPerezz) [reply .github/workflows/ci.yml:47]: The idea is that the `test` step uses the catched artifacts from the `build` step in the CI. I agree anyways we can have a sepparated job for the WASM-compat which will anyways not be re-used.

But I'd not split all the build from all the test in general due to the cache optimizations done for now with `nextest`.

As for the `target`. I completelly missed it! Thx!

Comment (CPerezz) [reply .github/workflows/ci.yml:45]: Fixed in [c5acdaf](https://github.com/privacy-scaling-explorations/halo2curves/pull/122/commits/c5acdafea008745b513f05dd169f72da71c28a22)

**Review by han0110 — APPROVED** [review]
LGTM! Just some final nits that can be ignored.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -32,18 +32,42 @@ env:
 concurrency:
   group: ${{ github.workflow }}-${{ github.head_ref || github.run_id }}
   cancel-in-progress: true
-  
 
 jobs:
+  compat:
+    if: github.event.pull_request.draft == false
+    name: Test
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply .github/workflows/ci.yml]: Ammended in 3f58eff

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -32,18 +32,42 @@ env:
...
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        target:
+          - wasm32-unknown-unknown
+          - wasm32-wasi
+        include:
+          - feature: default
+          - feature: bn256-table
+          - feature: derive_serde
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply .github/workflows/ci.yml]: Right! Fixed in 3f58eff

File: Cargo.toml [hunk 4]
```diff
@@ -35,11 +40,10 @@ serde = { version = "1.0", default-features = false, optional = true }
 serde_arrays = { version = "0.1.0", optional = true }
 hex = { version = "0.4", optional = true, default-features = false, features = ["alloc", "serde"] }
 blake2b_simd = "1"
-maybe-rayon = { version = "0.1.0", default-features = false }
+rayon = "1.8"
 
 [features]
-default = ["bits", "multicore"]
-multicore = ["maybe-rayon/threads"]
```
[the remark this answers is not in the document]
Comment (CPerezz) [reply Cargo.toml:42]: Ohh dammit!! 
Will update. Hopefully this will not affect the released crate.

**Pull request merged**